### PR TITLE
enable fused topK_softmax kernel for hip

### PR DESCRIPTION
**Pull request by divakar-amd**

### CODE REVIEW

**Comment by gshtras** [comment]
Measuring performance (latency) on the new fused implementation

**Comment by divakar-amd** [comment]
Tested on `rocm/pytorch-private:vllm_0.4.0_ROCm6.1_public_20.04_PyTorch2.3_tunable`
Enabling fused topK_softmax gives about 4-5% gain.
image

File: csrc/moe/topk_softmax_kernels.cu [hunk 1]
```diff
@@ -383,7 +390,7 @@ struct TopkConstants
 {
     static constexpr int ELTS_PER_LDG = BYTES_PER_LDG / sizeof(float);
     static_assert(EXPERTS / (ELTS_PER_LDG * WARP_SIZE) == 0 || EXPERTS % (ELTS_PER_LDG * WARP_SIZE) == 0, "");
-    static constexpr int VECs_PER_THREAD = std::max(1, EXPERTS / (ELTS_PER_LDG * WARP_SIZE));
+    static constexpr int VECs_PER_THREAD = MAX(1, EXPERTS / (ELTS_PER_LDG * WARP_SIZE));
```
Comment (gshtras) [inline csrc/moe/topk_softmax_kernels.cu:393]: Why though?

Comment (divakar-amd) [reply csrc/moe/topk_softmax_kernels.cu:393]: The hip compiler had a tough time using std::max to create a constexpr variable.
image

Comment (gshtras) [reply csrc/moe/topk_softmax_kernels.cu:393]: You're right. Such a shame.
We may want to submit a bug to hipclang for not complying with c++14 in /opt/rocm-6.1.0/lib/llvm/lib/clang/17/include/__clang_hip_math.h
I mean, it's only been 10 years, right?

**Review by hthangirala — COMMENTED** [review]
Looks good to me

**Review by hthangirala — APPROVED** [review]

**Pull request merged**